File: algorithms/2 - Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
			int c=0;
			ListNode t=null, ret=new ListNode(0);
			t=ret;
			while (l1!=null||l2!=null) {
				if(l1!=null){
					c+=l1.val;
					l1=l1.next;
				}
				if(l2!=null){
					c+=l2.val;
					l2=l2.next;
				}
				if(c>0){
				t.next=new ListNode((c>0)?c%10:0);
				c/=10;
				t=t.next;
				}else{
					t.next=new ListNode(0);
					t=t.next;
				}
			}
			if(c==1)t.next=new ListNode(1);
		return ret.next;
    }
}

File: algorithms/ruby/2 - Add Two Numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  head = ListNode.new 0
  p = l1
  q = l2
  curr = head
  carry = 0

  while p || q do
    x = p ? p.val : 0
    y = q ? q.val : 0
    sum = carry + x + y
    carry = sum / 10
    curr.next = ListNode.new sum % 10
    curr = curr.next
    p = p.next if p
    q = q.next if q
  end

  if carry > 0
    curr.next = ListNode.new carry
  end
  head.next
end
